In [91]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import LabelEncoder
%matplotlib inline 

**job** type of job (categorical: 'admin.','blue-collar','entrepreneur','housemaid','management','retired','self-employed','services','student','technician','unemployed','unknown')
<br> **marital** marital status (categorical: 'divorced','married','single','unknown'; note: 'divorced' means divorced or widowed)
<br> **education** (categorical: 'basic.4y','basic.6y','basic.9y','high.school','illiterate','professional.course','university.degree','unknown')	
<br> **default** has credit in default?
<br> **balance** average yearly balance	
<br> **housing** has housing loan?	
<br> **loan** has personal loan?	
<br> **contact** contact communication type (categorical: 'cellular','telephone')	
<br> **day_of_week** last contact day of the week	
<br> **y** if the client will subscribe a term deposit 

In [63]:
url = 'https://archive.ics.uci.edu/static/public/222/data.csv'
df = pd.read_csv(url, sep=',', engine='python')

print(df.head())

   age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married        NaN      no     1506     yes   no   
4   33           NaN   single        NaN      no        1      no   no   

  contact  day_of_week month  duration  campaign  pdays  previous poutcome   y  
0     NaN            5   may       261         1     -1         0      NaN  no  
1     NaN            5   may       151         1     -1         0      NaN  no  
2     NaN            5   may        76         1     -1         0      NaN  no  
3     NaN            5   may        92         1     -1         0      NaN  no  
4     NaN            5   may       198         1     -1         0      NaN  no  


In [65]:
print(df.columns)

Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day_of_week', 'month', 'duration', 'campaign',
       'pdays', 'previous', 'poutcome', 'y'],
      dtype='object')


In [67]:
feature_column_names = [
    'age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan',
    'contact', 'day_of_week', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome', 'y'
]

# If columns are not correct, try reading with header=None and assign columns manually
if df.columns[0] != 'age':
    df = pd.read_csv(url, sep=',', engine='python', header=None)
    df.columns = feature_column_names

df.head()


age           job  marital  education default  balance housing loan  \
0   58    management  married   tertiary      no     2143     yes   no   
1   44    technician   single  secondary      no       29     yes   no   
2   33  entrepreneur  married  secondary      no        2     yes  yes   
3   47   blue-collar  married        NaN      no     1506     yes   no   
4   33           NaN   single        NaN      no        1      no   no   

  contact  day_of_week month  duration  campaign  pdays  previous poutcome   y  
0     NaN            5   may       261         1     -1         0      NaN  no  
1     NaN            5   may       151         1     -1         0      NaN  no  
2     NaN            5   may        76         1     -1         0      NaN  no  
3     NaN            5   may        92         1     -1         0      NaN  no  
4     NaN            5   may       198         1     -1         0      NaN  no

In [71]:
df.columns

Index(['age', 'job', 'marital', 'education', 'default', 'balance', 'housing',
       'loan', 'contact', 'day_of_week', 'month', 'duration', 'campaign',
       'pdays', 'previous', 'poutcome', 'y'],
      dtype='object')

In [73]:
df.isnull().sum()/len(df)*100

age             0.000000
job             0.637013
marital         0.000000
education       4.107407
default         0.000000
balance         0.000000
housing         0.000000
loan            0.000000
contact        28.798301
day_of_week     0.000000
month           0.000000
duration        0.000000
campaign        0.000000
pdays           0.000000
previous        0.000000
poutcome       81.747805
y               0.000000
dtype: float64

In [83]:
#drop any rows with missing values
df = df.replace('unknown', pd.NA)

df = df.dropna()   

In [85]:
df.head()

age         job  marital  education default  balance housing loan  \
24060   33      admin.  married   tertiary      no      882      no   no   
24062   42      admin.   single  secondary      no     -247     yes  yes   
24064   33    services  married  secondary      no     3444     yes   no   
24072   36  management  married   tertiary      no     2415     yes   no   
24077   36  management  married   tertiary      no        0     yes   no   

         contact  day_of_week month  duration  campaign  pdays  previous  \
24060  telephone           21   oct        39         1    151         3   
24062  telephone           21   oct       519         1    166         1   
24064  telephone           21   oct       144         1     91         4   
24072  telephone           22   oct        73         1     86         4   
24077  telephone           23   oct       140         1    143         3   

      poutcome    y  
24060  failure   no  
24062    other  yes  
24064  failure  yes  
24072    other   no  
24077  failure  yes

In [87]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7842 entries, 24060 to 45210
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   age          7842 non-null   int64 
 1   job          7842 non-null   object
 2   marital      7842 non-null   object
 3   education    7842 non-null   object
 4   default      7842 non-null   object
 5   balance      7842 non-null   int64 
 6   housing      7842 non-null   object
 7   loan         7842 non-null   object
 8   contact      7842 non-null   object
 9   day_of_week  7842 non-null   int64 
 10  month        7842 non-null   object
 11  duration     7842 non-null   int64 
 12  campaign     7842 non-null   int64 
 13  pdays        7842 non-null   int64 
 14  previous     7842 non-null   int64 
 15  poutcome     7842 non-null   object
 16  y            7842 non-null   object
dtypes: int64(7), object(10)
memory usage: 1.1+ MB


In [103]:
#define features and targets
feature = [
    'age', 'job', 'marital', 'education', 'default', 'balance', 'housing', 'loan',
    'contact', 'day_of_week', 'month', 'duration', 'campaign', 'pdays', 'previous', 'poutcome'
]

target = 'y'

In [105]:
# separate features (X) and target (y)
X = df[feature]
y = df[target]

In [107]:
# encode categorical variables
label_encoders = {}
for column in X.select_dtypes(include=['object']).columns:
    le = LabelEncoder()
    X.loc[:, column] = le.fit_transform(X[column])
    label_encoders[column] = le

In [109]:
# encode the target variable
y = LabelEncoder().fit_transform(y)

In [111]:
# split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [113]:
#Decision Tree classifier
clf = DecisionTreeClassifier(random_state=42)
clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [115]:
# predict
y_pred = clf.predict(X_test)

In [117]:
# evaluate model
accuracy = accuracy_score(y_test, y_pred)
report = classification_report(y_test, y_pred)

In [119]:
print(f"Accuracy: {accuracy:.2f}")
print("Classification Report:")
print(report)

Accuracy: 0.79
Classification Report:
              precision    recall  f1-score   support

           0       0.87      0.86      0.86      1207
           1       0.54      0.56      0.55       362

    accuracy                           0.79      1569
   macro avg       0.70      0.71      0.71      1569
weighted avg       0.79      0.79      0.79      1569



## Results
- Class 0 (Not Subscribed):
    - Precision: 0.87
    - Recall: 0.86
    - F1-score: 0.86
>
- Class 1 (Subscribed):
    - Precision: 0.54
    - Recall: 0.56
    - F1-score: 0.55